10_rest/alextmz/cmd/puppy-server/main.go

```diff
@@ -0,0 +1,145 @@
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var (
+	args      = os.Args[1:]
+	flagfile  = kingpin.Flag("data", "JSON file to read").Short('d').Default("./test/valid-formatted-json.json").String()
+	flagport  = kingpin.Flag("port", "TCP port to listen on").Short('p').Default("7735").Uint16()
+	flagstore = kingpin.Flag("store", "Backing store to use").Short('s').Default("sync").Enum("map", "sync")
+
+	out io.Writer = os.Stdout
+
+	// shutdownhttp signals main() to... shutdown the http server
+	shutdownhttp = make(chan bool)
+
+	// syncoutput signals whoever interested that main() output is done
+	syncoutput = make(chan bool, 1)
+)
+
+func main() {
+	_, err := kingpin.CommandLine.Parse(args)
+	if err != nil {
+		fmt.Fprintf(out, "error parsing command line: %v\n", err)
+		return
+	}
+
+	jsonfile, err := os.Open(*flagfile)
+	if err != nil {
+		fmt.Fprintf(out, "error opening file: %v\n", err)
+		return
+	}
+	defer jsonfile.Close()
+
+	puppies, err := readfile(jsonfile)
+	if err != nil {
+		fmt.Fprintf(out, "error reading JSON file: %v\n", err)
+		return
+	}
+
+	var puppystore puppy.Storer
+
+	switch *flagstore {
+	case "map":
+		puppystore = store.NewMapStore()
+	case "sync":
+		puppystore = store.NewSyncStore()
+	}
+
+	n, err := storepuppies(puppystore, puppies)
+	if err != nil {
+		fmt.Fprintf(out, "error storing puppies: %v\n", err)
+		return
+	}
+
+	fmt.Fprintf(out, "Starting puppyserver with options:\n")
+	fmt.Fprintf(out, "file   = %s\nport  = %d\nstore = %s\n", *flagfile, *flagport, *flagstore)
+	fmt.Fprintf(out, "Loaded %d puppies.\n", n)
+	printpuppies(puppystore, n)
+
+	h := rest.HTTPHandler{Store: puppystore}
+	s := http.Server{
+		Addr:    ":" + strconv.Itoa(int(*flagport)),
+		Handler: h,
+	}
+
+	// synchttpshutdown forces main() to wait for http.Shutdown to complete.
+	synchttpshutdown := make(chan bool)
+
+	// if we get anything on shutdownhttp, signals the server to stop
+	// and makes main() wait for http.Shutdown to close up shop
+	go func() {
+		<-shutdownhttp
+		_ = s.Shutdown(context.Background())
+		close(synchttpshutdown)
+	}()
+	syncoutput <- true
+
+	err = s.ListenAndServe()
+	if err != nil && err == http.ErrServerClosed {
+		<-synchttpshutdown
+	}
+}
+
+// printpuppies print n puppies contained in the store s
+func printpuppies(s puppy.Storer, n int) {
+	for i := 1; i <= n; i++ {
+		p, err := s.ReadPuppy(i)
+		if err != nil {
+			fmt.Fprintf(out, "%v\n", err)
+			return
+		}
+
+		fmt.Fprintf(out, "Printing puppy id %d: %#v\n", i, p)
+	}
+}
+
+// storepuppies store all puppies contained in slice 'puppies'
+// into the store 'store', returning either (number of puppies stored, nil)
+// if there is no error or (0, error) if there was an error.
+func storepuppies(store puppy.Storer, puppies []puppy.Puppy) (int, error) {
+	n := 0
+
+	for _, v1 := range puppies {
+		v := v1
+
+		err := store.CreatePuppy(&v)
+		if err != nil {
+			return 0, err
+		}
+		n++
+	}
+
+	return n, nil
```

Comment:
`return len(puppies)-1, nil` should allow getting rid of `n` from this function

Reply:
Yep, actually `len(puppies)`. Changed!